scene/3d/mesh_instance.h

```diff
@@ -36,17 +36,37 @@
 #include "scene/resources/mesh.h"
 #include "scene/resources/skin.h"
 
+#include "core/local_vector.h"
+
 class MeshInstance : public GeometryInstance {
 
 	GDCLASS(MeshInstance, GeometryInstance);
 
+public:
+	enum SoftwareSkinningNormalMode {
+		DONT_TRANSFORM_NORMALS,
+		TRANSFORM_NORMALS,
+		ENSURE_CORRECT_NORMALS,
+	};
+
 protected:
 	Ref<Mesh> mesh;
+	Ref<Mesh> mesh_instance;
 	Ref<Skin> skin;
 	Ref<Skin> skin_internal;
 	Ref<SkinReference> skin_ref;
 	NodePath skeleton_path;
 
+	SoftwareSkinningNormalMode software_skinning_normal_mode;
+	bool software_skinning_local_buffer;
+	bool software_skinning_bones_ready;
+
+	struct S_SkinningSurfaceData {
+		PoolByteArray buffer;
+		PoolByteArray::Write buffer_write;
+	};
```

Comment:
I'm not so sure about storing the local buffer this way. `buffer_write` is never released to avoid any cow during skinning updates. It would be cleaner to store the data as a simple array of bytes instead, but it would mean making a variation of `mesh_surface_update_region` which takes a pointer instead of `PoolByteArray`.

Reply:
It's a bit of a workaround, but not that bad imo (providing it works ok). Using PoolArrays as the input to these visual server APIs may not have been ideal in retrospect, but writing alternative APIs is probably overkill and adds bug surface. Like you say people might have different opinions.

On the other hand I'm not familiar with the overall design, why PoolArrays are used for this kind of thing, maybe it exactly for thread safety (something else deciding to edit the mesh data in a different thread that we haven't anticipated), and by making our local copy we are violating this principle. This might be something we will need to consult with reduz on before going too deep down the rabbit hole.